File: backend/supabase_client.py
from supabase import create_client, Client
from dotenv import load_dotenv
import os
import json
from typing import Dict, Any, Optional, List

# Load environment variables
env_path = os.path.join(os.path.dirname(__file__), '..', '.env')
if os.path.isfile(env_path):
    load_dotenv(env_path)

class SupabaseClient:
    def __init__(self):
        self.supabase_url = os.getenv('SUPABASE_URL')
        self.supabase_key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
        
        if not self.supabase_url or not self.supabase_key:
            raise ValueError("Supabase URL and Service Role Key must be set in environment variables")
        
        self.client: Client = create_client(self.supabase_url, self.supabase_key)
    
    def get_client(self) -> Client:
        """Get the Supabase client instance"""
        return self.client
    
    # User Management
    async def get_user_by_email(self, email: str) -> Optional[Dict[str, Any]]:
        """Get user by email"""
        try:
            response = self.client.table('users').select('*').eq('email', email).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error getting user by email: {e}")
            return None
    
    async def get_user_by_id(self, user_id: int) -> Optional[Dict[str, Any]]:
        """Get user by ID"""
        try:
            response = self.client.table('users').select('*').eq('id', user_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error getting user by ID: {e}")
            return None
    
    async def create_user(self, user_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create new user"""
        try:
            response = self.client.table('users').insert(user_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error creating user: {e}")
            return None
    
    async def update_user(self, user_id: int, user_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update user"""
        try:
            response = self.client.table('users').update(user_data).eq('id', user_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error updating user: {e}")
            return None
    
    # Product Management
    async def get_products(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """Get products with optional filters"""
        try:
            query = self.client.table('products').select('*')
            
            if filters:
                if 'is_active' in filters:
                    query = query.eq('is_active', filters['is_active'])
                if 'seller_id' in filters:
                    query = query.eq('seller_id', filters['seller_id'])
                if 'category' in filters:
                    query = query.eq('category', filters['category'])
            
            response = query.execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting products: {e}")
            return []
    
    async def get_product_by_id(self, product_id: int) -> Optional[Dict[str, Any]]:
        """Get product by ID"""
        try:
            response = self.client.table('products').select('*').eq('id', product_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error getting product by ID: {e}")
            return None
    
    async def create_product(self, product_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create new product"""
        try:
            response = self.client.table('products').insert(product_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error creating product: {e}")
            return None
    
    async def update_product(self, product_id: int, product_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update product"""
        try:
            response = self.client.table('products').update(product_data).eq('id', product_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error updating product: {e}")
            return None
    
    # Cart Management
    async def get_cart_items(self, user_id: int) -> List[Dict[str, Any]]:
        """Get cart items for user"""
        try:
            response = self.client.table('cart').select('*').eq('user_id', user_id).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting cart items: {e}")
            return []
    
    async def add_to_cart(self, cart_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Add item to cart"""
        try:
            response = self.client.table('cart').insert(cart_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error adding to cart: {e}")
            return None
    
    async def update_cart_item(self, cart_id: int, user_id: int, cart_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update cart item"""
        try:
            response = self.client.table('cart').update(cart_data).eq('id', cart_id).eq('user_id', user_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error updating cart item: {e}")
            return None
    
    async def remove_cart_item(self, cart_id: int, user_id: int) -> bool:
        """Remove item from cart"""
        try:
            response = self.client.table('cart').delete().eq('id', cart_id).eq('user_id', user_id).execute()
            return len(response.data) > 0 if response.data else False
        except Exception as e:
            print(f"Error removing cart item: {e}")
            return False
    
    async def clear_cart(self, user_id: int) -> bool:
        """Clear all cart items for user"""
        try:
            response = self.client.table('cart').delete().eq('user_id', user_id).execute()
            return True
        except Exception as e:
            print(f"Error clearing cart: {e}")
            return False
    
    # Order Management
    async def create_order(self, order_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create new order"""
        try:
            response = self.client.table('orders').insert(order_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error creating order: {e}")
            return None
    
    async def get_orders_by_buyer(self, buyer_id: int) -> List[Dict[str, Any]]:
        """Get orders for buyer"""
        try:
            response = self.client.table('orders').select('*').eq('buyer_id', buyer_id).order('created_at', desc=True).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting buyer orders: {e}")
            return []
    
    async def get_orders_by_seller(self, seller_id: int) -> List[Dict[str, Any]]:
        """Get orders for seller"""
        try:
            response = self.client.table('orders').select('*').eq('seller_id', seller_id).order('created_at', desc=True).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting seller orders: {e}")
            return []
    
    async def update_order_status(self, order_id: int, status: str) -> Optional[Dict[str, Any]]:
        """Update order status"""
        try:
            response = self.client.table('orders').update({'status': status}).eq('id', order_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error updating order status: {e}")
            return None
    
    # Order Items
    async def create_order_items(self, order_items_data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Create order items"""
        try:
            response = self.client.table('order_items').insert(order_items_data).execute()
            return response.data or []
        except Exception as e:
            print(f"Error creating order items: {e}")
            return []
    
    async def get_order_items(self, order_id: int) -> List[Dict[str, Any]]:
        """Get order items"""
        try:
            response = self.client.table('order_items').select('*').eq('order_id', order_id).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting order items: {e}")
            return []
    
    # Product Stock Management
    async def get_product_stock(self, product_id: int, size: str = None, color: str = None) -> List[Dict[str, Any]]:
        """Get product stock"""
        try:
            query = self.client.table('product_size_stock').select('*').eq('product_id', product_id)
            
            if size:
                query = query.eq('size', size)
            if color:
                query = query.eq('color', color)
            
            response = query.execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting product stock: {e}")
            return []
    
    async def update_product_stock(self, stock_id: int, stock_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update product stock"""
        try:
            response = self.client.table('product_size_stock').update(stock_data).eq('id', stock_id).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error updating product stock: {e}")
            return None
    
    # Chat Management
    async def get_conversations(self, user_id: int, user_type: str = None) -> List[Dict[str, Any]]:
        """Get conversations for user"""
        try:
            if user_type == 'seller':
                response = self.client.table('chat_conversations').select('*').eq('seller_id', user_id).order('last_message_time', desc=True).execute()
            else:
                response = self.client.table('chat_conversations').select('*').eq('buyer_id', user_id).order('last_message_time', desc=True).execute()
            
            return response.data or []
        except Exception as e:
            print(f"Error getting conversations: {e}")
            return []
    
    async def get_messages(self, conversation_id: int) -> List[Dict[str, Any]]:
        """Get messages for conversation"""
        try:
            response = self.client.table('chat_messages').select('*').eq('conversation_id', conversation_id).order('created_at', asc=True).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting messages: {e}")
            return []
    
    async def send_message(self, message_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Send message"""
        try:
            response = self.client.table('chat_messages').insert(message_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error sending message: {e}")
            return None
    
    # Notifications
    async def get_notifications(self, user_id: int) -> List[Dict[str, Any]]:
        """Get notifications for user"""
        try:
            response = self.client.table('notifications').select('*').eq('user_id', user_id).order('created_at', desc=True).execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting notifications: {e}")
            return []
    
    async def create_notification(self, notification_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create notification"""
        try:
            response = self.client.table('notifications').insert(notification_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error creating notification: {e}")
            return None
    
    async def mark_notification_read(self, notification_id: int, user_id: int) -> bool:
        """Mark notification as read"""
        try:
            response = self.client.table('notifications').update({'is_read': True}).eq('id', notification_id).eq('user_id', user_id).execute()
            return len(response.data) > 0 if response.data else False
        except Exception as e:
            print(f"Error marking notification as read: {e}")
            return False
    
    # Applications
    async def create_application(self, application_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Create application"""
        try:
            response = self.client.table('applications').insert(application_data).execute()
            if response.data:
                return response.data[0]
            return None
        except Exception as e:
            print(f"Error creating application: {e}")
            return None
    
    async def get_applications(self, filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """Get applications with filters"""
        try:
            query = self.client.table('applications').select('*')
            
            if filters:
                if 'user_id' in filters:
                    query = query.eq('user_id', filters['user_id'])
                if 'application_type' in filters:
                    query = query.eq('application_type', filters['application_type'])
                if 'status' in filters:
                    query = query.eq('status', filters['status'])
            
            response = query.execute()
            return response.data or []
        except Exception as e:
            print(f"Error getting applications: {e}")
            return []

# Global instance
supabase_client = SupabaseClient()

File: mobile/lib/core/supabase/supabase_client.dart
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

class SupabaseService {
  static SupabaseClient? _client;
  static SupabaseService? _instance;

  SupabaseService._internal();

  factory SupabaseService() {
    _instance ??= SupabaseService._internal();
    return _instance!;
  }

  SupabaseClient get client {
    _client ??= Supabase.instance.client;
    return _client!;
  }

  static Future<void> initialize() async {
    await Supabase.initialize(
      url: dotenv.env['SUPABASE_URL'] ?? '',
      anonKey: dotenv.env['SUPABASE_ANON_KEY'] ?? '',
    );
  }

  // Authentication Methods
  Future<AuthResponse> signUp({
    required String email,
    required String password,
    required Map<String, dynamic> userMetadata,
  }) async {
    return await client.auth.signUp(
      email: email,
      password: password,
      data: userMetadata,
    );
  }

  Future<AuthResponse> signIn({
    required String email,
    required String password,
  }) async {
    return await client.auth.signInWithPassword(
      email: email,
      password: password,
    );
  }

  Future<void> signOut() async {
    await client.auth.signOut();
  }

  User? get currentUser => client.auth.currentUser;

  // Database Methods
  Future<PostgrestList<T>> getAll<T>({
    required String tableName,
    List<String> selectColumns = const ['*'],
    Map<String, dynamic>? filters,
    String? orderBy,
    bool ascending = true,
  }) async {
    var query = client.from(tableName).select(selectColumns.join(', '));

    if (filters != null) {
      filters.forEach((key, value) {
        query = query.eq(key, value);
      });
    }

    if (orderBy != null) {
      query = query.order(orderBy, ascending: ascending);
    }

    return await query as PostgrestList<T>;
  }

  Future<PostgrestList<T>> getByField<T>({
    required String tableName,
    required String fieldName,
    required dynamic value,
    List<String> selectColumns = const ['*'],
  }) async {
    return await client
        .from(tableName)
        .select(selectColumns.join(', '))
        .eq(fieldName, value) as PostgrestList<T>;
  }

  Future<PostgrestList<T>> getById<T>({
    required String tableName,
    required int id,
    List<String> selectColumns = const ['*'],
  }) async {
    return await client
        .from(tableName)
        .select(selectColumns.join(', '))
        .eq('id', id) as PostgrestList<T>;
  }

  Future<PostgrestResponse> insert({
    required String tableName,
    required Map<String, dynamic> data,
    List<String> selectColumns = const ['*'],
  }) async {
    return await client
        .from(tableName)
        .insert(data)
        .select(selectColumns.join(', '));
  }

  Future<PostgrestResponse> update({
    required String tableName,
    required int id,
    required Map<String, dynamic> data,
    List<String> selectColumns = const ['*'],
  }) async {
    return await client
        .from(tableName)
        .update(data)
        .eq('id', id)
        .select(selectColumns.join(', '));
  }

  Future<PostgrestResponse> delete({
    required String tableName,
    required int id,
  }) async {
    return await client.from(tableName).delete().eq('id', id);
  }

  // User specific methods
  Future<PostgrestList<Map<String, dynamic>>> getUserByEmail(String email) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'users',
      fieldName: 'email',
      value: email,
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getUserById(int userId) async {
    return await getById<Map<String, dynamic>>(
      tableName: 'users',
      id: userId,
    );
  }

  Future<PostgrestResponse> createUser(Map<String, dynamic> userData) async {
    return await insert(
      tableName: 'users',
      data: userData,
    );
  }

  Future<PostgrestResponse> updateUser(int userId, Map<String, dynamic> userData) async {
    return await update(
      tableName: 'users',
      id: userId,
      data: userData,
    );
  }

  // Product specific methods
  Future<PostgrestList<Map<String, dynamic>>> getProducts({
    bool? isActive,
    int? sellerId,
    String? category,
  }) async {
    Map<String, dynamic>? filters = {};
    if (isActive != null) filters['is_active'] = isActive;
    if (sellerId != null) filters['seller_id'] = sellerId;
    if (category != null) filters['category'] = category;

    return await getAll<Map<String, dynamic>>(
      tableName: 'products',
      filters: filters,
      orderBy: 'created_at',
      ascending: false,
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getProductById(int productId) async {
    return await getById<Map<String, dynamic>>(
      tableName: 'products',
      id: productId,
    );
  }

  Future<PostgrestResponse> createProduct(Map<String, dynamic> productData) async {
    return await insert(
      tableName: 'products',
      data: productData,
    );
  }

  Future<PostgrestResponse> updateProduct(int productId, Map<String, dynamic> productData) async {
    return await update(
      tableName: 'products',
      id: productId,
      data: productData,
    );
  }

  // Cart specific methods
  Future<PostgrestList<Map<String, dynamic>>> getCartItems(int userId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'cart',
      fieldName: 'user_id',
      value: userId,
    );
  }

  Future<PostgrestResponse> addToCart(Map<String, dynamic> cartData) async {
    return await insert(
      tableName: 'cart',
      data: cartData,
    );
  }

  Future<PostgrestResponse> updateCartItem(int cartId, int userId, Map<String, dynamic> cartData) async {
    return await client
        .from('cart')
        .update(cartData)
        .eq('id', cartId)
        .eq('user_id', userId)
        .select();
  }

  Future<PostgrestResponse> removeCartItem(int cartId, int userId) async {
    return await client
        .from('cart')
        .delete()
        .eq('id', cartId)
        .eq('user_id', userId)
        .select();
  }

  Future<PostgrestResponse> clearCart(int userId) async {
    return await client
        .from('cart')
        .delete()
        .eq('user_id', userId)
        .select();
  }

  // Order specific methods
  Future<PostgrestResponse> createOrder(Map<String, dynamic> orderData) async {
    return await insert(
      tableName: 'orders',
      data: orderData,
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getOrdersByBuyer(int buyerId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'orders',
      fieldName: 'buyer_id',
      value: buyerId,
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getOrdersBySeller(int sellerId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'orders',
      fieldName: 'seller_id',
      value: sellerId,
    );
  }

  Future<PostgrestResponse> updateOrderStatus(int orderId, String status) async {
    return await update(
      tableName: 'orders',
      id: orderId,
      data: {'status': status},
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getOrderItems(int orderId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'order_items',
      fieldName: 'order_id',
      value: orderId,
    );
  }

  Future<PostgrestResponse> createOrderItems(List<Map<String, dynamic>> orderItemsData) async {
    return await client
        .from('order_items')
        .insert(orderItemsData)
        .select();
  }

  // Notification specific methods
  Future<PostgrestList<Map<String, dynamic>>> getNotifications(int userId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'notifications',
      fieldName: 'user_id',
      value: userId,
    );
  }

  Future<PostgrestResponse> createNotification(Map<String, dynamic> notificationData) async {
    return await insert(
      tableName: 'notifications',
      data: notificationData,
    );
  }

  Future<PostgrestResponse> markNotificationRead(int notificationId, int userId) async {
    return await client
        .from('notifications')
        .update({'is_read': true})
        .eq('id', notificationId)
        .eq('user_id', userId)
        .select();
  }

  // Chat specific methods
  Future<PostgrestList<Map<String, dynamic>>> getConversations(int userId, String userType) async {
    String fieldName = userType == 'seller' ? 'seller_id' : 'buyer_id';
    return await client
        .from('chat_conversations')
        .select('*')
        .eq(fieldName, userId)
        .order('last_message_time', ascending: false);
  }

  Future<PostgrestList<Map<String, dynamic>>> getMessages(int conversationId) async {
    return await getByField<Map<String, dynamic>>(
      tableName: 'chat_messages',
      fieldName: 'conversation_id',
      value: conversationId,
    );
  }

  Future<PostgrestResponse> sendMessage(Map<String, dynamic> messageData) async {
    return await insert(
      tableName: 'chat_messages',
      data: messageData,
    );
  }

  // Application specific methods
  Future<PostgrestResponse> createApplication(Map<String, dynamic> applicationData) async {
    return await insert(
      tableName: 'applications',
      data: applicationData,
    );
  }

  Future<PostgrestList<Map<String, dynamic>>> getApplications({
    int? userId,
    String? applicationType,
    String? status,
  }) async {
    Map<String, dynamic>? filters = {};
    if (userId != null) filters['user_id'] = userId;
    if (applicationType != null) filters['application_type'] = applicationType;
    if (status != null) filters['status'] = status;

    return await getAll<Map<String, dynamic>>(
      tableName: 'applications',
      filters: filters,
    );
  }

  // Stock specific methods
  Future<PostgrestList<Map<String, dynamic>>> getProductStock(
    int productId, {
    String? size,
    String? color,
  }) async {
    var query = client.from('product_size_stock').select('*').eq('product_id', productId);
    
    if (size != null) query = query.eq('size', size);
    if (color != null) query = query.eq('color', color);

    return await query as PostgrestList<Map<String, dynamic>>;
  }

  Future<PostgrestResponse> updateProductStock(int stockId, Map<String, dynamic> stockData) async {
    return await update(
      tableName: 'product_size_stock',
      id: stockId,
      data: stockData,
    );
  }
}
